<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Package for Demultiplexing Cell Hashing Data • cellhashR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Package for Demultiplexing Cell Hashing Data">
<meta property="og:description" content="The goal of this package is to provide tools for QC and demultiplexing of cell hashing data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">cellhashR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/V01-QC-example.html">Data/Normalization QC</a>
    </li>
    <li>
      <a href="articles/V02-BFF-example.html">Demultiplexing with Bimodal Flexible Fitting (BFF)</a>
    </li>
    <li>
      <a href="articles/V03-Benchmark-example.html">Benchmarking Demultiplexing Algorithms with cellhashR</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BimberLab/cellhashR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div class="section level1">
<div class="page-header"><h1 id="cellhashr">cellhashR<a class="anchor" aria-label="anchor" href="#cellhashr"></a>
</h1></div>
<p>An R package designed to demultiplex cell hashing data. <a href="https://bimberlab.github.io/cellhashR/" class="external-link">Please see our documentation for more detail</a>.</p>
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ul>
<li><a href="#overview">Overview</a></li>
<li><a href="#example">Example Usage</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#issues">Known Issues</a></li>
<li><a href="#developers">Development Guidelines</a></li>
</ul>
<div class="section level3">
<h3 id="overview">
<a name="overview">Overview</a><a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>Cell hashing is a method that allows sample multiplexing or super-loading within single-cell RNA-seq platforms, such as 10x genomics, originally developed at New York Genome Center in collaboration with the Satija lab. <a href="https://cite-seq.com/cell-hashing/" class="external-link">See here for more detail on the technique</a>. The general idea is that cells are labeled with a staining reagent (such as an antibody) tagged with a short nucleotide barcode. Other staining methods have been published, such as the lipid-based Multi-Seq (<a href="https://www.ncbi.nlm.nih.gov/pubmed/31209384" class="external-link">https://www.ncbi.nlm.nih.gov/pubmed/31209384</a>). In all methods, the hashtag oligo/barcode is sequenced in parallel with cellular mRNA, creating a separate cell hashing library. After sequencing, the cell barcode and hashing index are parsed using tools like Cite-seq-Count (<a href="https://github.com/Hoohm/CITE-seq-Count" class="external-link">https://github.com/Hoohm/CITE-seq-Count</a>), creating a count matrix with the total hash tag counts per cell.</p>
<p>Once the count matrix is created, an algorithm must be used to demultiplex cells and assign them to hash tags (i.e. sample). This is where cellhashR comes in. This package provides several functions: - Quality control reports for the cell hashing library, covering read counts and normalization. Think <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FASTQC</a>, except for cell hashing data. - A single interface to run one or more demutiplexing algorithms, including the novel demultiplexing algorithms BFF_raw and BFF_cluster. Each algorithm has pros and cons, and will perform better or worse under certain conditions (though in our experience, of the algorithms we have tested, the BFF algorithms work most consistently and under the widest variety of conditions). If you select multiple algorithms (our default workflow), cellhashR will score cells using the consensus call from the set. Various QC summaries are produced during this process as well, if debugging is needed. In addition to the BFF demultiplexing algorithms, other algorithms that can be run from cellhashR include: - <a href="https://github.com/CHPGenetics/GMM-Demux" class="external-link">GMM-Demux</a> - <a href="https://github.com/klarman-cell-observatory/demuxEM" class="external-link">demuxEM</a> <a href="#demuxEM">(see extra requirements below)</a> - <a href="https://github.com/huklein/demuxmix" class="external-link">demuxmix</a> <a href="#demuxmix">(see extra requirements below)</a> - <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq" class="external-link">deMULTIplex</a> - <a href="https://satijalab.org/seurat/v3.1/hashing_vignette.html" class="external-link">HTODemux from Seurat</a> - <a href="https://github.com/MarioniLab/DropletUtils" class="external-link">hashedDrops from DropletUtils</a> - The workflow produces a unified table with the results of each caller and the consensus call. Final QC plots and summaries are created.</p>
<p>Each step of the workflow can either be run interactively in R (through the terminal or RStudio), or it can be executed as a pipeline that runs all commands and creates the call table and an HTML report.</p>
<p><a href="https://bimberlab.github.io/cellhashR/articles/V01-QC-example.html" class="external-link">Click here to view an example QC report</a></p>
</div>
<div class="section level3">
<h3 id="consensus-calling">
<a name="example">Consensus Calling</a><a class="anchor" aria-label="anchor" href="#consensus-calling"></a>
</h3>
<p>In addition to allowing one to run multiple demuliplexing algorithms to compare results, cellhashR can generate a consensus call based on those scores. This can be useful, since some algorithms will perform better or worse under some conditions. This is automatically built into the dataframe returned by GenerateCellHashingCalls(). Some additional parameters that might be worth considering are: - There are separate arguments for ‘methods’ (i.e. which algorithms will be run), and ‘methodsForConsensus’, which determined the subset that will be used for the consensus call. - majorityConsensusThreshold: This applies to calculating a consensus call when multiple algorithms are used. If NULL, then all non-negative calls must agree or that cell is marked discordant. If non-NULL, then the number of algorithms returning the top call is divided by the total number of non-negative calls. If this ratio is above the majorityConsensusThreshold, that value is selected. For example, when majorityConsensusThreshold=0.6 and the calls are: HTO-1,HTO-1,Negative,HTO-2, then 2/3 calls are for HTO-1, giving 0.66. This is greater than the majorityConsensusThreshold of 0.6, so HTO-1 is returned. This can be useful for situations where most algorithms agree, but a single caller fails. - callerDisagreementThreshold: If provided, the agreement rate will be calculated between each caller and the simple majority call, ignoring discordant and no-call cells. If any caller has an disagreement rate above this threshold, it will be dropped and the consensus call re-calculated. The general idea is to drop a caller that is systematically discordant.</p>
</div>
<div class="section level3">
<h3 id="example-usage">
<a name="example">Example Usage</a><a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h3>
<p>Below are the primary functions of cellhashR needed to QC and score hashing data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example 1: parse CITE-seq-Count output, printing QC</span></span>
<span><span class="va">barcodeData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProcessCountMatrix.html">ProcessCountMatrix</a></span><span class="op">(</span>rawCountData <span class="op">=</span> <span class="st">'myCountDir/umi_count'</span>, minCountPerCell <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Example 2: parse CITE-seq-Count output, providing a barcode whitelist. </span></span>
<span><span class="va">barcodeData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProcessCountMatrix.html">ProcessCountMatrix</a></span><span class="op">(</span>rawCountData <span class="op">=</span> <span class="st">'myCountDir/umi_count'</span>, minCountPerCell <span class="op">=</span> <span class="fl">5</span>, barcodeWhitelist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'HTO-1'</span>, <span class="st">'HTO-2'</span>, <span class="st">'HTO-3'</span>, <span class="st">'HTO-4'</span>, <span class="st">'HTO-6'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create QC plots of barcode normalization</span></span>
<span><span class="fu"><a href="reference/PlotNormalizationQC.html">PlotNormalizationQC</a></span><span class="op">(</span><span class="va">barcodeData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate the final cell hashing calls</span></span>
<span><span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/GenerateCellHashingCalls.html">GenerateCellHashingCalls</a></span><span class="op">(</span>barcodeMatrix <span class="op">=</span> <span class="va">barcodeData</span>, methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'multiseq'</span>, <span class="st">'htodemux'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect negative cells:</span></span>
<span><span class="fu"><a href="reference/SummarizeCellsByClassification.html">SummarizeCellsByClassification</a></span><span class="op">(</span>calls <span class="op">=</span> <span class="va">calls</span>, barcodeMatrix <span class="op">=</span> <span class="va">barcodeData</span><span class="op">)</span></span></code></pre></div>
<p>Or export/save a template RMarkdown file outlining the default workflow, which can be run interactively or headlessly as part of a pipeline:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/GetExampleMarkdown.html">GetExampleMarkdown</a></span><span class="op">(</span>dest <span class="op">=</span> <span class="st">'cellhashR_template.rmd'</span><span class="op">)</span></span></code></pre></div>
<p>Finally, the workflow can be executed using this wrapper around the Rmarkdown, producing a TSV of calls and HTML QC report:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/CallAndGenerateReport.html">CallAndGenerateReport</a></span><span class="op">(</span>rawCountData <span class="op">=</span> <span class="st">'myCountDir/umi_count'</span>, reportFile <span class="op">=</span> <span class="st">'report.html'</span>, callFile <span class="op">=</span> <span class="st">'calls.txt'</span>, barcodeWhitelist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'HTO-1'</span>, <span class="st">'HTO-2'</span>, <span class="st">'HTO-3'</span><span class="op">)</span>, title <span class="op">=</span> <span class="st">'Cell Hashing For Experiment 1'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="installation">
<a name="installation">Installation</a><a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<pre class="{r}"><code><span><span class="co"># Make sure to update your Rprofile to include Bioconductor repos, such as adding this line to ~/.Rprofile:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/eval.html" class="external-link">local</a></span><span class="op">(</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Latest version:</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">'bimberlab/cellhashR'</span>, ref <span class="op">=</span> <span class="st">'master'</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span>, upgrade <span class="op">=</span> <span class="st">'always'</span><span class="op">)</span></span></code></pre>
<p>Pre-packaged Docker images with all needed dependencies installed can be found on our <a href="https://github.com/orgs/BimberLab/packages/container/package/cellhashr" class="external-link">GitHub Packages page</a>. We recommend using a specific release, which you can do using tags:</p>
<pre><code>docker pull ghcr.io/bimberlab/cellhashr:latest</code></pre>
</div>
<div class="section level3">
<h3 id="known-issues">
<a name="issues">Known Issues</a><a class="anchor" aria-label="anchor" href="#known-issues"></a>
</h3>
<p>If you receive an error along the lines of:</p>
<pre><code><span><span class="st">"ERROR; return code from pthread_create() is 22\n"</span></span></code></pre>
<p>Please manually install preprocessCore with threading disabled:</p>
<pre><code><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'bmbolstad/preprocessCore'</span>, dependencies <span class="op">=</span> <span class="cn">T</span>, upgrade <span class="op">=</span> <span class="st">'always'</span>, configure.args <span class="op">=</span> <span class="st">'--disable-threading'</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="demuxemdemuxmix">
<a name="demuxEM">demuxEM/demuxmix</a><a class="anchor" aria-label="anchor" href="#demuxemdemuxmix"></a>
</h3>
<p>Unlike the other algorithms, which just require the HTO count matrix, demuxEM and demuxmix also require the 10x h5 gene expression counts. This can be supplied as follows. This example runs BFF and demuxEM:</p>
<pre><code><span>  <span class="va">rawData</span> <span class="op">&lt;-</span> <span class="st">'../testdata/438-21-GEX/umi_count'</span></span>
<span>  <span class="va">h5File</span> <span class="op">&lt;-</span> <span class="st">'../testdata/438-21-GEX/438-21-raw_feature_bc_matrix.h5'</span></span>
<span>  <span class="va">barcodeMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProcessCountMatrix.html">ProcessCountMatrix</a></span><span class="op">(</span>rawCountData <span class="op">=</span> <span class="va">rawData</span>, barcodeWhitelist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MS-11'</span>, <span class="st">'MS-12'</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/GenerateCellHashingCalls.html">GenerateCellHashingCalls</a></span><span class="op">(</span>barcodeMatrix <span class="op">=</span> <span class="va">barcodeMatrix</span>, methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'bff_cluster'</span>, <span class="st">'demuxem'</span><span class="op">)</span>, demuxem.rawFeatureMatrixH5 <span class="op">=</span> <span class="va">h5File</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="development-guidelines">
<a name="developers">Development Guidelines</a><a class="anchor" aria-label="anchor" href="#development-guidelines"></a>
</h3>
<ul>
<li><p>New development should occur on a branch, and go through a Pull Request before merging into the master branch. <a href="https://guides.github.com/introduction/flow/" class="external-link">See here for information on the pull request workflow</a>. Ideally PRs would be reviewed by another person. For the PR, please review the set of changed files carefully to make sure you are only merging the changes you intend.</p></li>
<li><p>New functions should have <a href="https://kbroman.org/pkg_primer/pages/docs.html" class="external-link">Roxygen2 documentation</a>.</p></li>
<li><p>As part of each PR, you should run ‘devtools::document()’ to update documentation and include these changes with your commits.</p></li>
<li><p>It is a good idea to run ‘R CMD check’ locally to make sure your changes will pass. <a href="http://r-pkgs.had.co.nz/check.html" class="external-link">See here for more information</a></p></li>
<li><p>Code should only be merged after the build and tests pass. The master branch should always be stable.</p></li>
<li><p>New features should ideally have at least a basic test (see <a href="http://r-pkgs.had.co.nz/tests.html" class="external-link">R testthat</a>). There is existing test data in ./tests/testdata. This can be expanded, but please be conscious about file size and try to reuse data across tests if appropriate.</p></li>
</ul>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/BimberLab/cellhashR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/BimberLab/cellhashR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing cellhashR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>The package maintainer <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://zenodo.org/badge/latestdoi/317274382" class="external-link"><img src="https://zenodo.org/badge/317274382.svg" alt="DOI"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
