<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Package for Demultiplexing Cell Hashing Data • cellhashR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Package for Demultiplexing Cell Hashing Data">
<meta property="og:description" content="The goal of this package is to provide tools for QC and demultiplexing of cell hashing data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">cellhashR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BimberLab/cellhashR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="https://github.com/BimberLab/cellhashR/workflows/R%20Build%20and%20Checks/badge.svg" alt="R Build and Checks"></p>
<div id="cellhashr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#cellhashr" class="anchor"></a>cellhashR</h1></div>
<p>An R package designed to demultiplex cell hashing data. <a href="https://bimberlab.github.io/cellhashR/">Please see our documentation for more detail</a>.</p>
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ul>
<li><a href="#overview">Overview</a></li>
<li><a href="#example">Example Usage</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#developers">Development Guidelines</a></li>
</ul>
<div id="overview" class="section level3">
<h3 class="hasAnchor">
<a href="#overview" class="anchor"></a><a name="overview">Overview</a>
</h3>
<p>Cell hashing is a method that allows sample multiplexing or super-loading within single-cell RNA-seq platforms, such as 10x genomics, originally developed at New York Genome Center in collaboration with the Satija lab. <a href="https://cite-seq.com/cell-hashing/">See here for more detail on the technique</a>. The general idea is that cells are labeled with a staining reagent (such as an antibody) tagged with a short nucleotide barcode. Other staining methods have been published, such as the lipid-based Multi-Seq (<a href="https://www.ncbi.nlm.nih.gov/pubmed/31209384">https://www.ncbi.nlm.nih.gov/pubmed/31209384</a>). In all methods, the hashtag oligo/barcode is sequenced in parallel with cellular mRNA, creating a separate cell hashing library. After sequencing, the cell barcode and hashing index are parsed using tools like Cite-seq-Count (<a href="https://github.com/Hoohm/CITE-seq-Count">https://github.com/Hoohm/CITE-seq-Count</a>), creating a count matrix with the total hash tag counts per cell.</p>
<p>Once the count matrix is created, an algorithm must be used to demultiplex cells and assign them to hash tags (i.e. sample). This is where cellhashR comes in. This package provides several functions:</p>
<ul>
<li>Quality control reports for the cell hashing library, covering read counts and normalization. Think <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a>, except for cell hashing data.</li>
<li>A single interface to run one or more demutiplexing algorithms. Multiple algorithms have been published to demultiplex and classify cells, including <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq">deMULTIplex</a> and <a href="https://satijalab.org/seurat/v3.1/hashing_vignette.html">HTODemux from Seurat</a>. Each has pros and cons, and will perform better or worse under certain conditions. If you select multiple algorithms (our default workflow), cellhashR will score cells using the consensus call from the set. Various QC summaries are produced during this process as well, if debugging is needed.</li>
<li>The workflow produces a unified table with the results of each caller and the consensus call. Final QC plots and summaries are created.</li>
</ul>
<p>Each step of the workflow can either be run interactively in R (through the terminal or RStudio), or it can be executed as a pipeline that runs all commands and creates the call table and an HTML report.</p>
<p><a href="https://htmlpreview.github.io/?https://github.com/BimberLab/cellhashR/blob/master/examples/cellhashR.html">Click here to download an example HTML report</a></p>
</div>
<div id="example-usage" class="section level3">
<h3 class="hasAnchor">
<a href="#example-usage" class="anchor"></a><a name="example">Example Usage</a>
</h3>
<p>Below are the primary functions of cellhashR needed to QC and score hashing data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example 1: parse CITE-seq-Count output, printing QC</span>
<span class="va">barcodeData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProcessCountMatrix.html">ProcessCountMatrix</a></span><span class="op">(</span>rawCountData <span class="op">=</span> <span class="st">'myCountDir/umi_count'</span>, minCountPerCell <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="co"># Example 2: parse CITE-seq-Count output, providing a barcode whitelist. </span>
<span class="va">barcodeData</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ProcessCountMatrix.html">ProcessCountMatrix</a></span><span class="op">(</span>rawCountData <span class="op">=</span> <span class="st">'myCountDir/umi_count'</span>, minCountPerCell <span class="op">=</span> <span class="fl">5</span>, barcodeWhitelist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'HTO-1'</span>, <span class="st">'HTO-2'</span>, <span class="st">'HTO-3'</span>, <span class="st">'HTO-4'</span>, <span class="st">'HTO-6'</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Create QC plots of barcode normalization</span>
<span class="fu"><a href="reference/PlotNormalizationQC.html">PlotNormalizationQC</a></span><span class="op">(</span><span class="va">barcodeData</span><span class="op">)</span>

<span class="co"># Generate the final cell hashing calls</span>
<span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/GenerateCellHashingCalls.html">GenerateCellHashingCalls</a></span><span class="op">(</span>barcodeMatrix <span class="op">=</span> <span class="va">barcodeData</span>, methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'multiseq'</span>, <span class="st">'htodemux'</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Inspect negative cells:</span>
<span class="fu"><a href="reference/SummarizeCellsByClassification.html">SummarizeCellsByClassification</a></span><span class="op">(</span>calls <span class="op">=</span> <span class="va">calls</span>, barcodeMatrix <span class="op">=</span> <span class="va">barcodeData</span><span class="op">)</span>
</code></pre></div>
<p>Or export/save a template RMarkdown file outlining the default workflow, which can be run interactively or headlessly as part of a pipeline:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/GetExampleMarkdown.html">GetExampleMarkdown</a></span><span class="op">(</span>dest <span class="op">=</span> <span class="st">'cellhashR_template.rmd'</span><span class="op">)</span></code></pre></div>
<p>Finally, the workflow can be executed using this wrapper around the Rmarkdown, producing a TSV of calls and HTML QC report:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/CallAndGenerateReport.html">CallAndGenerateReport</a></span><span class="op">(</span>rawCountData <span class="op">=</span> <span class="st">'myCountDir/umi_count'</span>, reportFile <span class="op">=</span> <span class="st">'report.html'</span>, callFile <span class="op">=</span> <span class="st">'calls.txt'</span>, barcodeWhitelist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'HTO-1'</span>, <span class="st">'HTO-2'</span>, <span class="st">'HTO-3'</span><span class="op">)</span>, title <span class="op">=</span> <span class="st">'Cell Hashing For Experiment 1'</span><span class="op">)</span></code></pre></div>
</div>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a><a name="installation">Installation</a>
</h3>
<pre class="{r}"><code><a href="https://rdrr.io/r/base/eval.html"># Make sure to update your Rprofile to include Bioconductor repos, such as adding this line to ~/.Rprofile:
local({options(repos = BiocManager::repositories())})

#Latest version:
devtools::install_github(repo = 'bimberlab/cellhashR', ref = 'master', dependencies = TRUE, upgrade = 'always')</a></code></pre>
<p>Pre-packaged Docker images with all needed dependencies installed can be found on our <a href="https://hub.docker.com/r/bimberlab/oosap">github repository</a>:</p>
<pre><code>docker pull ghcr.io/bimberlab/cellhashR:latest</code></pre>
</div>
<div id="development-guidelines" class="section level3">
<h3 class="hasAnchor">
<a href="#development-guidelines" class="anchor"></a><a name="developers">Development Guidelines</a>
</h3>
<ul>
<li><p>New development should occur on a branch, and go through a Pull Request before merging into the master branch. <a href="https://guides.github.com/introduction/flow/">See here for information on the pull request workflow</a>. Ideally PRs would be reviewed by another person. For the PR, please review the set of changed files carefully to make sure you are only merging the changes you intend.</p></li>
<li><p>New functions should have <a href="https://kbroman.org/pkg_primer/pages/docs.html">Roxygen2 documentation</a>.</p></li>
<li><p>As part of each PR, you should run ‘devtools::document()’ to update documentation and include these changes with your commits.</p></li>
<li><p>It is a good idea to run ‘R CMD check’ locally to make sure your changes will pass. <a href="http://r-pkgs.had.co.nz/check.html">See here for more information</a></p></li>
<li><p>Code should only be merged after the build and tests pass. The master branch should always be stable.</p></li>
<li><p>New features should ideally have at least a basic test (see <a href="http://r-pkgs.had.co.nz/tests.html">R testthat</a>). There is existing test data in ./tests/testdata. This can be expanded, but please be conscious about file size and try to reuse data across tests if appropriate.</p></li>
</ul>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/BimberLab/cellhashR/">https://​github.com/​BimberLab/​cellhashR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/BimberLab/cellhashR/issues">https://​github.com/​BimberLab/​cellhashR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>The package maintainer <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
